resources/qml/Widgets/SingleSettingTextField.qml

```diff
@@ -0,0 +1,188 @@
+UM.TextField
+{
+    id: control
+    property alias settingName: propertyProvider.key
+
+    // If true, all extruders will have "settingName" property updated.
+    // The displayed value will be read from the extruder with index "defaultExtruderIndex" instead of the machine.
+    property bool updateAllExtruders: false
+    // This is only used if updateAllExtruders == true
+    property int defaultExtruderIndex: 0
+
+    // Resolving the value in the textField.
+    Binding
+    {
+        target: control
+        property: "text"
+
+        value:
+        {
+            if (control.activeFocus)
+            {
+                // This stops the text being reformatted as you edit. For example "10.1" -Edit-> "10." -Auto Format-> "10.0".
+                return control.text
+            }
+
+            if (( propertyProvider.properties.resolve != "None" &&  propertyProvider.properties.resolve) && ( propertyProvider.properties.stackLevels[0] != 0) && ( propertyProvider.properties.stackLevels[0] != 1))
+            {
+                // We have a resolve function. Indicates that the setting is not settable per extruder and that
+                // we have to choose between the resolved value (default) and the global value
+                // (if user has explicitly set this).
+                return base.resolve
+            }
+
+            return propertyProvider.properties.value
+        }
+
+    }
+
+    property UM.SettingPropertyProvider propertyProvider: UM.SettingPropertyProvider
+    {
+        id: propertyProvider
+        watchedProperties: ["value", "validationState",  "resolve"]
+        containerStackId: updateAllExtruders ? Cura.ExtruderManager.extruderIds[defaultExtruderIndex] : Cura.MachineManager.activeMachine.id
+    }
+
+    Connections
+    {
+        target: propertyProvider
+        function onContainerStackChanged()
+        {
+            updateTimer.restart()
+        }
+        function onIsValueUsedChanged()
+        {
+            updateTimer.restart()
+        }
+    }
+
+    // Restart update timer right after releasing a key. This stops lag while typing, but you still get warning and error
+    // textfield styling while typing.
+    Keys.onReleased: updateTimer.restart()
+    // Forces formatting when you finish editing "10.1" -Edit-> "10." -Focus Change-> "10"
+    onActiveFocusChanged: updateTimer.restart()
+
+    // Updates to the setting are delayed by interval. This stops lag caused by calling the
+    // parseValueUpdateSetting() function being called repeatedly while changing the text value.
+    Timer
+    {
+        id: updateTimer
+        interval: 50
+        repeat: false
+        onTriggered: parseValueUpdateSetting()
+    }
+
+    function parseValueUpdateSetting()
+    {
+        if (propertyProvider.properties.value == text || (parseFloat(propertyProvider.properties.value) == parseFloat(text) && !isNan(parseFloat(text))))
```

Comment:
```suggestion
        if (propertyProvider.properties.value === text || (parseFloat(propertyProvider.properties.value) === parseFloat(text) && !isNan(parseFloat(text))))
```

Reply:
Also as 
```
$ Number.NaN === Number.NaN
> false
```
it is not needed to do the `!isNan(..)` part of the comparison; as we are already checking for equality of `parseFloat(propertyProvider.properties.value)` and `parseFloat(text)` before the AND. As `NaN != NaN` is not possible after that equality check for any of the parts to be `NaN`.

```suggestion
        if (propertyProvider.properties.value === text || (parseFloat(propertyProvider.properties.value) === parseFloat(text)))
```